### ISSUE TITLE
@doc redefinition warning does not appear when doc is split between function heads

### ISSUE THREAD

**Issue by novaugust** [elixir-lang/elixir#7817]
### Environment

* Elixir & Erlang/OTP versions (elixir --version): 
```
Erlang/OTP 20 [erts-9.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.6.4 (compiled with OTP 20)
```

* Operating system: `OSX 10.13.5`

### Current behavior

No compile-time warning appears for doc redefinitions if they are split by function heads.

This generates no warnings, but obviously should:

```ex
defmodule A do
  @doc "doc"
  def foo(:a), do: :a
  @doc "redefined"
  def foo(:b), do: :b
end
```

### Expected behavior

Should see an error ala

```
warning: redefining @doc attribute previously set at line ~x~
```

**Comment by josevalim** [comment]
My concern is that someone may generate functions with default docs and someone may want to replace those default docs.

We warn when we have two `@doc` in a row exactly because it is 100% guaranteed it is a user error. But the scenario you describe can be intentional. Maybe we should look if it is a bodiless clause? So we warn here:

```elixir
  @doc "doc"
  def foo(:a), do: :a
  @doc "redefined"
  def foo(:b), do: :b
```

But we wouldn't warn for this:

```elixir
  @doc "doc"
  def foo(:a), do: :a
  @doc "redefined"
  def foo(foo)
```

As you are clearly showing intent to overwrite docs? Thoughts?

**Comment by novaugust** [comment]
I found this and refactored everything into one doc, and the documentation was all the better for it. If each function head truly deserves its own documentation, then it's probably better off as a completely separate function. It gets hard to reason about the function when each clause has its own notes scattered about. If it's implementation notes-per-head, then you're better off with comments anyways.

And actually, I did change it to have a bodiless head to help with the documentation, leaving each implementation without a `@doc`, but that's not totally relevant I reckon =)

tl;dr, I would've appreciated having the warnings, for sure.

Re: generation, dang, that's tricky. There's no way to know if the doc was generated by a macro / defoverrideable eh?
